### PR TITLE
Implemented reset_index for multiindex and multilevel indexes

### LINKED ISSUE #2530: Implement level and multiindex cases of reset_index
Currently some parts of `reset_index` default to Pandas: https://github.com/modin-project/modin/blob/master/modin/backends/pandas/query_compiler.py#L564

### PR DESCRIPTION
**Pull request by gshimansky**
## What do these changes do?

- [x] commit message follows format outlined [here](https://modin.readthedocs.io/en/latest/CONTRIBUTING.html)
- [x] passes `flake8 modin`
- [x] passes `black --check modin`
- [x] signed commit with `git commit -s` 
- [x] Resolves #2530  
- [x] tests added and passing

### CODE REVIEW

File: modin/pandas/test/dataframe/test_indexing.py [hunk 1]
```diff
@@ -844,6 +844,107 @@ def test_reset_index(data):
...
+    modin_df.reset_index(drop=True, inplace=True)
+    pandas_df.reset_index(drop=True, inplace=True)
+    df_equals(modin_df, pandas_df)
+
+    modin_df = pd.DataFrame(test_data_values[0])
+    pandas_df = pandas.DataFrame(test_data_values[0])
+    modin_df.index.name = pandas_df.index.name = "NEW_NAME"
+    df_equals(modin_df.reset_index(drop=False), pandas_df.reset_index(drop=False))
+
+
```
Comment (dchigarev) [inline modin/pandas/test/dataframe/test_indexing.py:947]:
```suggestion
@pytest.mark.parametrize(
    "level",
    [None, 0, 1, 2, [2, 0], [1, 1]],
)
@pytest.mark.parametrize("col_level", [0, 1, 2])
@pytest.mark.parametrize("col_fill", ["", "name", 5, 5.5, None])
def test_reset_index_with_multi_index_no_drop(level, col_level, col_fill):
    data = test_data_values[0]
    index = generate_multiindex(len(data[list(data.keys())[0]]), nlevels=3)
    columns = generate_multiindex(len(data.keys()), nlevels=3)

    md_df, pd_df = create_test_dfs(data, index=index)
    eval_general(
        *create_test_dfs(data, index=index, columns=columns),
        lambda df: df.reset_index(
            level=level, drop=False, col_level=col_level, col_fill=col_fill
        ),
    )


@pytest.mark.parametrize(
    "level",
    [None, 0, 1, 2, [2, 0], [1, 1]],
)
def test_reset_index_with_multi_index_drop(level):
    data = test_data_values[0]
    index = generate_multiindex(len(data[list(data.keys())[0]]), nlevels=3)

    md_df, pd_df = create_test_dfs(data, index=index)
    eval_general(
        *create_test_dfs(data, index=index),
        lambda df: df.reset_index(level=level, drop=True),
    )


@pytest.mark.parametrize("is_multiindex", [True, False])
@pytest.mark.parametrize("is_name_multiindex", [True, False])
def test_reset_index_with_named_index(is_multiindex, is_name_multiindex):
    data = test_data_values[0]
    index = (
        generate_multiindex(len(data[list(data.keys())[0]]), nlevels=3)
        if is_multiindex
        else pandas.Index(np.arange(len(data[list(data.keys())[0]]), 0, -1))
    )
    columns = generate_multiindex(len(data.keys()), nlevels=3)
    index_names = (
        [(f"level{i}", 0) for i in range(index.nlevels)]
        if is_name_multiindex
        else [f"level{i}" for i in range(index.nlevels)]
    )

    index.names = index_names
    eval_general(
        *create_test_dfs(data, index=index, columns=columns),
        lambda df: df.reset_index(drop=False),
    )


```
previous version of testing contains more than 900 and didn't cover all cases, tests from suggestion contains about ~100 tests and revealed many of failing cases:
```
pytest test_indexing.py -k "test_reset_index"
========= 37 failed, 65 passed, 118 deselected, 105 warnings in 40.44s =========
```

File: modin/backends/pandas/query_compiler.py [hunk 2]
```diff
@@ -561,21 +561,48 @@ def reset_index(self, **kwargs):
...
+                mi_frame = self.index.to_frame()
+                if level is not None:
+                    if not isinstance(level, (tuple, list)):
+                        level = [level]
+                    level = [self.index._get_level_number(lev) for lev in level]
+                    uniq_sorted_level = sorted(set(level))
+                    if len(level) < self.index.nlevels:
+                        new_index = self.index.droplevel(uniq_sorted_level)
+                    mi_frame = mi_frame.iloc[:, uniq_sorted_level]
+                if self.columns.nlevels > 1 and mi_frame.columns.nlevels == 1:
```
Comment (dchigarev) [inline modin/backends/pandas/query_compiler.py:579]: condition should be something like:
```python
if self.columns.nlevels > mi_frame.columns.nlevels
```
 and the logic in this if-branch should also be changed, because it's going to processes situations like this incorrectly:
```python
>>> df
                           a     b
                           0  1  0  1
                           z  z  z  z
(level-0, 0) (level-0, 1)
a            0             0  0  0  0
             1             1  1  1  1
b            0             2  2  2  2
             1             3  3  3  3
             2             4  4  4  4
>>> df.index
MultiIndex([('a', 0),
            ('a', 1),
            ('b', 0),
            ('b', 1),
            ('b', 2)],
           names=[('level-0', 0), ('level-0', 1)])
>>> df.columns
MultiIndex([('a', 0, 'z'),
            ('a', 1, 'z'),
            ('b', 0, 'z'),
            ('b', 1, 'z')],
           )
>>> df.reset_index(drop=False)
  level-0     a     b
        0  1  0  1  0  1
              z  z  z  z
0       a  0  0  0  0  0
1       a  1  1  1  1  1
2       b  0  2  2  2  2
3       b  1  3  3  3  3
4       b  2  4  4  4  4
```

File: modin/pandas/test/dataframe/test_indexing.py (hunk 1 above)
Comment (gshimansky) [reply modin/pandas/test/dataframe/test_indexing.py:947]: Ok I agree that some multiindex configurations were missing in the test, but I don't agree that with dropping a lot of parameter combinations. The fast that they pass the test doesn't mean that they aren't there for a reason. I would rather add a case with multilevel column names (unequal to number of column levels in the frame, BTW it also makes sense to add a case with more levels in multiindex than in the frame, not just a case with less levels) to my test than switch to your proposal.

File: modin/backends/pandas/query_compiler.py (hunk 2 above)
Comment (gshimansky) [reply modin/backends/pandas/query_compiler.py]: I updated tests and functions with tuples in level names. Number of test cases only grew up because of it, but I think that all parameter combinations are necessary.

File: modin/backends/pandas/query_compiler.py [hunk 3]
```diff
@@ -561,21 +561,92 @@ def reset_index(self, **kwargs):
...
+                                    col_level
+                                    + (
+                                        len(mi_frame_level)
+                                        if isinstance(mi_frame_level, tuple)
+                                        else 1
+                                    ),
+                                    self.columns.nlevels,
+                                )
+                            ]
+                        )
```
Comment (dchigarev) [inline modin/backends/pandas/query_compiler.py:625]:
```suggestion
                        top_level = [col_fill] * col_level
                        middle_level = (
                            list(mi_frame_level)
                            if isinstance(mi_frame_level, tuple)
                            else [mi_frame_level]
                        )
                        bottom_level = [col_fill] * (
                            self.columns.nlevels - (col_level + len(middle_level))
                        )
                        item = tuple(top_level + middle_level + bottom_level)
```
This `item` construction looks heavy to read, can we change it somehow?

BTW, why do we need to convert `mi_frame_level` to the list when it's tuple?

**Comment by dchigarev** [comment]
This implementation works incorrect in case of None names of multi-index. This implementation inserts columns with `None` names inside the frame when `drop=False`, when pandas replaces `None` names to the `"level-0", "level-1", ...`

We should probably add a check at the beginning of `reset_index` if all index names are `None`, and if it is - rename them similar to pandas.
<details><summary> Reproducer </summary>


```python
import numpy as np
import modin.pandas as pd
import pandas
from modin.pandas.test.utils import df_equals, generate_multiindex

index = generate_multiindex(5)
index.names = [None, None]

columns = generate_multiindex(5)
data = {f"col{i}": np.arange(5) for i in range(5)}

md_df = pd.DataFrame(data, index=index, columns=columns)
pd_df = pandas.DataFrame(data, index=index, columns=columns)

md_res, pd_res = md_df.reset_index(drop=False), pd_df.reset_index(drop=False)
df_equals(md_res, pd_res) # AssertionError: columns are different

```



</details>

File: modin/pandas/test/dataframe/test_indexing.py [hunk 4]
```diff
@@ -886,6 +887,116 @@ def test_reset_index(data):
     df_equals(modin_df_cp, pd_df_cp)
 
 
+@pytest.mark.parametrize("data", test_data_values, ids=test_data_keys)
```
Comment (dchigarev) [inline modin/pandas/test/dataframe/test_indexing.py:890]: this PR adds 4900 new tests only for `reset_index` (previously number of tests for the whole DataFrame API was about ~2200), so the time of running `test_indexing.py` increased a lot, here is a time comparison of executing `test_indexing.py` for a few of our GH-actions job
```
                                            master    #2558
test-ubuntu (engine python, python 3.6)     49s       17m 33s
test-ubuntu (engine dask, python 3.7)       1m 56s    1h 0m 20s
``` 
we should decrease the amount of tests for `reset_index` somehow

File: modin/backends/pandas/query_compiler.py [hunk 5]
```diff
@@ -561,21 +561,92 @@ def reset_index(self, **kwargs):
...
+                # tupleize_cols is True by default
+                mi_frame.columns = pandas.Index(self.index.names, tupleize_cols=False)
+
+                if level is not None:
+                    mi_frame = mi_frame.iloc[:, uniq_sorted_level]
+                if isinstance(self.columns, pandas.MultiIndex):
+                    items = []
+                    if col_fill is None:
+                        # Initialize col_fill if it is None.
+                        # This is some weird undocumented Pandas behavior to take first
```
Comment (dchigarev) [inline modin/backends/pandas/query_compiler.py:590]: Maybe instead trying to emulate pandas behaviour in terms of constructing new index and columns, we should try to create an empty pandas frame with the modins index and columns, apply `reset_index` on it with all passed parameters and then extract proper `new_index` and `new_columns` 
```python
dummy_df = pandas.DataFrame(index=md_df.index[:1], columns=md_df.columns).reset_index(*args, **kwargs)
new_index, new_columns = dummy_df.index, dummy_df.columns
```
but we should bench this approach and compare its time with the current implementation

File: modin/pandas/test/dataframe/test_indexing.py (hunk 4 above)
Comment (gshimansky) [reply modin/pandas/test/dataframe/test_indexing.py:890]: You don't have to run all these tests simultaneously. When I run them on 56 cores system with `-n 56` I get 20k tests executed in 2 minutes.

File: modin/backends/pandas/query_compiler.py (hunk 3 above)
Comment (gshimansky) [reply modin/backends/pandas/query_compiler.py:625]: Ok this looks like a good suggestion. And `mi_frame_level` has to be converted into a list for concatenation with `[col_fill]` lists before and after. Tuples cannot be concatenated.

File: modin/backends/pandas/query_compiler.py (hunk 5 above)
Comment (gshimansky) [reply modin/backends/pandas/query_compiler.py:595]: This will likely depend on the parameter combinations and we currently don't have any `reset_index` benchmarks, not to mention with multiple different parameter combinations. BTW why do you want to slice index as `[:1]`? Won't it make reset index concatenate only 1st row from the index levels into the frame?

Comment (dchigarev) [reply modin/backends/pandas/query_compiler.py:595]: We need it to get new proper columns, while building new columns `reset_index` considers only old level names, which is not dependent on the index length. We can even take 0 elements from the original index since we need only meta information without values.

<details><summary> Code example </summary>


```python
import numpy as np
import modin.pandas as pd
import pandas
from modin.pandas.test.utils import df_equals, generate_multiindex, create_test_dfs

index = generate_multiindex(5)
columns = generate_multiindex(5)
data = {col: np.arange(5) for col in columns}

md_df, pd_df = create_test_dfs(data, index=index, columns=columns)
dummy_df = pandas.DataFrame(index=md_df.index[:0], columns=md_df.columns)

md_res, pd_res = dummy_df.reset_index(drop=False), pd_df.reset_index(drop=False)
print(md_res.columns.equals(pd_res.columns))  # True
```



</details>

File: modin/pandas/test/dataframe/test_indexing.py (hunk 4 above)
Comment (dchigarev) [reply modin/pandas/test/dataframe/test_indexing.py:890]: But it will be just impossible to run all of the `reset_index` tests for external contributors on their home machine. I've ran these tests with 24 cores and it took me about 9 minutes to execute them all (vs few seconds previously). Imagine how frustrating it will be to rerun these tests locally again and again while changing/debugging `reset_index` functionality, while not having 56 cores. I strongly disagree about adding that big amount of tests. I believe that we can cover the same level of functionality by the significantly less amount of tests.

File: modin/backends/pandas/query_compiler.py (hunk 5 above)
Comment (dchigarev) [reply modin/backends/pandas/query_compiler.py:595]: I've tried this approach on top of the copy of this branch in [this commit](https://github.com/dchigarev/modin/commit/ab0fb15e84a36b5d68775e97f7a432cc8674d2a0) and all of the tests passed.

File: modin/pandas/test/dataframe/test_indexing.py [hunk 6]
```diff
@@ -886,6 +887,132 @@ def test_reset_index(data):
...
+    if none_in_index_names is True:
+        index.names = [None] * len(index.names)
+    elif none_in_index_names:
+        names_list = list(index.names)
+        start_index = 0 if none_in_index_names == "mixed_1st_None" else 1
+        names_list[start_index::2] = [None] * len(names_list[start_index::2])
+        index.names = names_list
+
+    modin_df = pd.DataFrame(data, index=index, columns=columns)
+    pandas_df = pandas.DataFrame(data, index=index, columns=columns)
```
Comment (dchigarev) [inline modin/pandas/test/dataframe/test_indexing.py:954]: Notice that when we're specifying custom columns at the DataFrame constructor it does not replace labels of the passed dict, but doing reindex:
```python
>>> data = {f"col{i}": np.arange(5) for i in range(4)}
>>> columns = ["a", "b", "col0", "c"]
>>> pandas.DataFrame(data, columns=columns)
     a    b  col0    c
0  NaN  NaN     0  NaN
1  NaN  NaN     1  NaN
2  NaN  NaN     2  NaN
3  NaN  NaN     3  NaN
4  NaN  NaN     4  NaN
>>> columns[2] = "d"
>>> pandas.DataFrame(data, columns=columns)
Empty DataFrame
Columns: [a, b, d, c]
Index: []
```
So, in terms of this test, it seems that we're always creating an empty frame.

We should either explicitly set index columns after DataFrame was created `df.columns = columns`, or convert data-dict to the 2D array to discard column labels.

File: modin/backends/pandas/query_compiler.py (hunk 5 above)
Comment (gshimansky) [reply modin/backends/pandas/query_compiler.py:595]: Yes I understand that you should get correct columns with `md_df.index[:1]` but my question was about assigning `dummy_df.index` to `new_index` in 2nd line of your code. You are supposed to concatenate all index rows into the frame columns.

File: modin/pandas/test/dataframe/test_indexing.py (hunk 6 above)
Comment (gshimansky) [reply modin/pandas/test/dataframe/test_indexing.py:956]: Thank you, I wasn't aware of this behavior of DataFrame constructor. I fixed test to modify `data` with generated column keys and it seems to work as intended now. A good thing that this behavior was debugged already on a "simple" dataframe which I later removed as redundant. So no new bugs were found after I fixed the test.

File: modin/backends/pandas/query_compiler.py (hunk 5 above)
Comment (dchigarev) [reply modin/backends/pandas/query_compiler.py:595]: Yes, you're right. We can't retrieve `new_index` from `dummy_df`.

File: .github/workflows/ci.yml [hunk 7]
```diff
@@ -90,10 +90,10 @@ jobs:
       - run: sudo apt update && sudo apt install -y libhdf5-dev
       - name: Api tests
         shell: bash -l {0}
-        run: python -m pytest modin/pandas/test/test_api.py
+        run: python -m pytest -n 2 modin/pandas/test/test_api.py
```
Comment (devin-petersohn) [inline .github/workflows/ci.yml:93]: Test changes should be in a separate PR.

**Review by devin-petersohn — COMMENTED** [review]
I think there's a simpler way to handle this, no?

We should be able to do a map on the first column partition. If we cannot estimate what the new row Index will be, we can just extract it from the first partitions. Then we can reassign it back out if there are more column partitions.

File: .github/workflows/ci.yml (hunk 7 above)
Comment (gshimansky) [reply .github/workflows/ci.yml:93]: Ok I will remove this commit because it didn't help anyways. Windows tests always fail in parallel because of Ray bug https://github.com/ray-project/ray/issues/13213

File: modin/engines/base/frame/data.py [hunk 8]
```diff
@@ -546,31 +546,46 @@ def from_labels(self) -> "BasePandasFrame":
...
-            new_column_names = pandas.Index(
+            level_names = [
+                self.index.names[0]
+                if self.index.names[0] is not None
+                else "index"
+                if "index" not in self.columns
+                else "level_{}".format(0)
+            ]
+
+        if self.columns.nlevels > 1:
```
Comment (devin-petersohn) [inline modin/engines/base/frame/data.py:569]: Is an exception thrown if we do not have the correct labels?

File: modin/engines/base/frame/data.py [hunk 9]
```diff
@@ -546,31 +546,46 @@ def from_labels(self) -> "BasePandasFrame":
...
+            new_column_names = pandas.MultiIndex.from_tuples(
+                # Set level names on the 1st columns level. This is how reset_index works
+                # when col_level is not specified.
+                # Fill up empty level names with empty string. This is how reset_index works
```
Comment (devin-petersohn) [inline modin/engines/base/frame/data.py:573]: The comment here is a little redundant, could you clean it up somewhat?

File: modin/engines/base/frame/data.py [hunk 10]
```diff
@@ -546,31 +546,46 @@ def from_labels(self) -> "BasePandasFrame":
...
-            new_column_names = pandas.Index(
-                [
-                    self.index.names[i]
-                    if self.index.names[i] is not None
-                    else "level_{}".format(i)
-                    for i in range(len(self.index.names))
-                ]
-            )
-            new_columns = new_column_names.append(self.columns)
+        # We will also use the `new_column_names` in the calculation of the internal metadata, so this is a
```
Comment (devin-petersohn) [inline modin/engines/base/frame/data.py:551]: `new_column_names` is no longer assigned here. Please update.

**Review by devin-petersohn — COMMENTED** [review]
One main question on this. Overall it looks pretty good.

File: modin/engines/base/frame/data.py (hunk 9 above)
Comment (gshimansky) [reply modin/engines/base/frame/data.py:573]: It is necessary to ensure unique names for all levels or pandas later throws an exception in `df.reset_index()` in `from_labels_executor`. It is also essential to create `pandas.MultiIndex` object with the same number of levels as `self.column.nlevels`. It could be possible to use fake names, e.g. "level_{#}" and always later replace them with correct ones in query_compiler, but creating a fake `pandas.MultiIndex` object is almost the same code as creating a correct one. It doesn't complicate the code too much and makes `from_labels` operator to output the most common case when `col_level` and `col_fill` are not specified.

File: modin/engines/base/frame/data.py (hunk 8 above)
Comment (gshimansky) [reply modin/engines/base/frame/data.py:569]: It is necessary to ensure unique names for all levels or pandas later throws an exception in `df.reset_index()` in `from_labels_executor`. It is also essential to create `pandas.MultiIndex` object with the same number of levels as `self.column.nlevels`. It could be possible to use fake names, e.g. "level_{#}" and always later replace them with correct ones in query_compiler, but creating a fake `pandas.MultiIndex` object is almost the same code as creating a correct one. It doesn't complicate the code too much and makes `from_labels` operator to output the most common case when `col_level` and `col_fill` are not specified.

File: modin/engines/base/frame/data.py (hunk 9 above)
Comment (gshimansky) [reply modin/engines/base/frame/data.py]: Yes. I updated it.

File: modin/engines/base/frame/data.py (hunk 10 above)
Comment (gshimansky) [reply modin/engines/base/frame/data.py]: Ok.

File: modin/engines/base/frame/data.py (hunk 8 above)
Comment (gshimansky) [reply modin/engines/base/frame/data.py:568]: Also, when creating fake column names, it is always a danger that some of those names already exist in the DataFrame. I figured that I can always forge test column names that would present a name collision, so it is better to use a safe approach repeating Pandas logic.

Comment (devin-petersohn) [reply modin/engines/base/frame/data.py:568]: Makes sense, thanks!

**Review by devin-petersohn — APPROVED** [review]
Thanks @gshimansky, LGTM!
